### PR TITLE
Add "Update Branch" button to Pull Requests

### LINKED ISSUE #9750: "Update Branch" button on PR
Related #8659 

Similar to GitHub when a branch on a PR is out of sync, then an "Updated Branch" button could be used to merge commits from upstream into PR branch (assuming no merge conflicts)

### PR DESCRIPTION
**Pull request by 6543**
close #9750

## Look

Bildschirmfoto zu 2020-01-17 02-00-04

Bildschirmfoto zu 2020-01-17 01-59-33

## Changes
 * [x] UI
   - [x] CSS
   - [x] Show Warning
   - [x] Display update button if user has right to update
 * [x] add go functions
   - [x] use existing merge code
   - [x] add a check to make sure there is a difference betwen the branches
   - [x] remove dublicated code
 * [x] Add & Fix Tests
   - [x] add a new PR witch can be updated
   - [x] test if update needed
   - [x] update
   - [x] test result

## Future Improvements:
 * different Update Modis (https://github.com/isaacs/github/issues/1419)
 * Update IsUserAllowedToUpdate to allow maintainer of baseRepo to change branch on headRepo
  -> PR-Author must allow this access (need migration to add this option)
 * imprufe Default Message for merges (#9393) -> ignore merge commits from baseBranch

### CODE REVIEW

**Comment by 6543** [comment]
## Button:
Bildschirmfoto zu 2020-01-16 02-04-01


## Result:
Bildschirmfoto zu 2020-01-16 02-07-35

**Comment by lunny** [comment]
I would like to put it like below because the right space of the merge button will be used by manually merge hints.

图片

File: routers/repo/pull.go [hunk 1]
```diff
@@ -587,6 +594,71 @@ func ViewPullFiles(ctx *context.Context) {
...
+
+	allowedUpdate, err := pull_service.IsUserAllowedToUpdate(issue.PullRequest, ctx.User)
+	if err != nil {
+		ctx.ServerError("IsUserAllowedToMerge", err)
+		return
+	}
+
+	// ToDo: add check if maintainers are allowed to change branch ... (need migration & co)
+	if !allowedUpdate {
+		ctx.Flash.Error(ctx.Tr("repo.pulls.update_not_allowed"))
```
Comment (lafriks) [inline routers/repo/pull.go:636]: Missing text for such translation key

**Comment by lafriks** [comment]
> I would like to put it like below because the right space of the merge button will be used by manually merge hints.
> 
> 图片

I think it should be placed in PR status box as new section

**Comment by lafriks** [comment]
I meant it would be better to place info about branch not being up to date after lin 159

File: templates/repo/issue/view_content/pull.tmpl [hunk 2]
```diff
@@ -269,6 +269,24 @@
 									</div>
 								</div>
 							</div>
+							{{if and .Divergence (gt .Divergence.Behind 0)}}
+							  <div class="ui divider"></div>
+								<div class="ui two horizontal list maxwidth">
+									<div class="item text gray">
+										<i class="icon icon-octicon"><span class="octicon octicon-alert">&nbsp;&nbsp;&nbsp;<b>{{$.i18n.Tr "repo.pulls.outdated_with_base_branch"}}</b></span></i>
```
Comment (techknowlogick) [inline templates/repo/issue/view_content/pull.tmpl]: instead of `<b>` please use `<strong>`

**Comment by 6543** [comment]
## Current Look:

Bildschirmfoto zu 2020-01-16 23-51-37

**Comment by 6543** [comment]
I added 18eb4951cf261673858c88a2db058ae634cc9bed because the page reoladed to early some times and then, the ref did not exist (while updating in that moment)

File: templates/repo/issue/view_content/pull.tmpl (hunk 2 above)
Comment (6543) [reply templates/repo/issue/view_content/pull.tmpl]: is in next UI commit

**Comment by 6543** [comment]
## Current Look:
Bildschirmfoto zu 2020-01-17 00-03-57

File: templates/repo/issue/view_content/pull.tmpl [hunk 3]
```diff
@@ -157,6 +157,24 @@
 							{{$.i18n.Tr (printf "repo.signing.wont_sign.%s" .WontSignReason) }}
 						</div>
 					{{end}}
+					{{if and .Divergence (gt .Divergence.Behind 0)}}
+						<div class="ui divider"></div>
```
Comment (lafriks) [inline templates/repo/issue/view_content/pull.tmpl]: No need for this divider:
```suggestion
```

Comment (6543) [reply templates/repo/issue/view_content/pull.tmpl]: removed (in next push)

**Comment by 6543** [comment]
Bildschirmfoto zu 2020-01-17 00-14-53

**Comment by lafriks** [comment]
I meant like this (more compact and imho looks better):
image

**Comment by 6543** [comment]
@lafriks me too but @lunny wrote about merge information?!?

**Comment by 6543** [comment]
> lunny: "the right space of the merge button will be used by manually merge hints."
@lafriks that msg ...

**Comment by lafriks** [comment]
ok: here is updated style:
https://github.com/go-gitea/gitea/commit/547c41e7468e8e50fd717f261f9127dfe7d04a75

image

File: options/locale/locale_en-US.ini [hunk 4]
```diff
@@ -1082,6 +1082,10 @@ pulls.open_unmerged_pull_exists = `You cannot perform a reopen operation because
 pulls.status_checking = Some checks are pending
 pulls.status_checks_success = All checks were successful
 pulls.status_checks_error = Some checks failed
+pulls.update_branch = Update branch
+pulls.update_branch_success = Update branch was successful
```
Comment (lafriks) [inline options/locale/locale_en-US.ini]:
```suggestion
pulls.update_branch_success = Branch update was successful
```

File: routers/repo/pull.go [hunk 5]
```diff
@@ -587,6 +601,74 @@ func ViewPullFiles(ctx *context.Context) {
 	ctx.HTML(200, tplPullFiles)
 }
 
+// UpdatePullRequest merge master into PR
+func UpdatePullRequest(ctx *context.Context) {
+
```
Comment (lafriks) [inline routers/repo/pull.go]:
```suggestion
```

File: routers/repo/pull.go [hunk 6]
```diff
@@ -587,6 +601,74 @@ func ViewPullFiles(ctx *context.Context) {
...
+	issue := checkPullInfo(ctx)
+	if ctx.Written() {
+		return
+	}
+	if issue.IsClosed {
+		ctx.NotFound("MergePullRequest", nil)
+		return
+	}
+
```
Comment (lafriks) [inline routers/repo/pull.go]:
```suggestion
```

File: routers/routes/routes.go [hunk 7]
```diff
@@ -855,6 +855,7 @@ func RegisterRoutes(m *macaron.Macaron) {
 			m.Get(".patch", repo.DownloadPullPatch)
 			m.Get("/commits", context.RepoRef(), repo.ViewPullCommits)
 			m.Post("/merge", context.RepoMustNotBeArchived(), reqRepoPullsWriter, bindIgnErr(auth.MergePullRequestForm{}), repo.MergePullRequest)
+			m.Post("/update", ignSignInAndCsrf, repo.UpdatePullRequest)
```
Comment (lafriks) [inline routers/routes/routes.go]: Why `ignSignInAndCsrf`?

**Review by lafriks — CHANGES_REQUESTED** [review]

File: routers/routes/routes.go (hunk 7 above)
Comment (6543) [reply routers/routes/routes.go]: to make sure user is loged in

Comment (6543) [reply routers/routes/routes.go]: @lafriks done

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
Please update screenshot for blog post ;)

**Comment by 6543** [comment]
sure

**Comment by 6543** [comment]
done

**Comment by 6543** [comment]
@lunny done

**Review by lunny — APPROVED** [review]

**Comment by 6543** [comment]
I'm ready if you are :rocket:

File: routers/repo/pull.go [hunk 8]
```diff
@@ -342,8 +343,21 @@ func PrepareViewPullInfo(ctx *context.Context, issue *models.Issue) *git.Compare
...
 
+	divergence, err := pull_service.GetDiverging(pull)
+	if err != nil {
+		ctx.ServerError("GetDiverging", err)
+		return nil
+	}
+	ctx.Data["Divergence"] = divergence
+	allowUpdate, err := pull_service.IsUserAllowedToUpdate(pull, ctx.User)
+	if err != nil {
+		ctx.ServerError("GetDiverging", err)
```
Comment (guillep2k) [inline routers/repo/pull.go:354]:
```suggestion
		ctx.ServerError("IsUserAllowedToUpdate", err)
```

File: routers/repo/pull.go [hunk 9]
```diff
@@ -587,6 +601,72 @@ func ViewPullFiles(ctx *context.Context) {
...
+
+	// ToDo: add check if maintainers are allowed to change branch ... (need migration & co)
+	if !allowedUpdate {
+		ctx.Flash.Error(ctx.Tr("repo.pulls.update_not_allowed"))
+		ctx.Redirect(ctx.Repo.RepoLink + "/pulls/" + com.ToStr(issue.Index))
+		return
+	}
+
+	// default merge commit message
+	message := fmt.Sprintf("Merge branch '%s' into %s", issue.PullRequest.BaseBranch, issue.PullRequest.HeadBranch)
```
Comment (guillep2k) [inline routers/repo/pull.go:642]: This should be configurable via .ini or via Crowdin.

**Review by guillep2k — APPROVED** [review]
Great work! 🚀

**Comment by guillep2k** [comment]
LGTM, but only two nit that can be addressed now if we'd like to.

**Comment by lunny** [comment]
@6543 I think @guillep2k 's comment https://github.com/go-gitea/gitea/pull/9784#discussion_r367765097 should be followed.

**Comment by 6543** [comment]
@guillep2k it was to late :( ...

but now we can talk if we should localice merge commit ...

File: routers/repo/pull.go (hunk 9 above)
Comment (lunny) [reply routers/repo/pull.go:642]: But for this one, sometimes, we use English commit messages even if I'm using a Chinese User Interface.

Comment (guillep2k) [reply routers/repo/pull.go:642]: We can use `app.ini` like we did with closing keywords and WIP...

**Comment by GrayedFox** [comment]
Your friends from [isaacs/github#1419](https://github.com/isaacs/github/issues/1419) say hi 👋

**Pull request merged**